File: src/test/scala/com/danylchuk/swiftlearner/nn/backprop/BackpropClassifierTest.scala
package com.danylchuk.swiftlearner.nn.backprop

import com.danylchuk.swiftlearner.MemoryTesting
import com.danylchuk.swiftlearner.data.{FisherIris, Mnist}
import com.typesafe.scalalogging.LazyLogging
import org.specs2.mutable.Specification


class BackpropClassifierTest extends Specification with LazyLogging with MemoryTesting {
  "BackpropClassifier" should {
    "classify the flowers from the Fisher Iris dataset" >> {
      val (trainingSet, testSet) = FisherIris.trainingAndTestDataFloat()

      // Normalize the input values to speed up learning
      def normalize(x: Float): Float = (x - 25) / 25

      val start = System.currentTimeMillis
      val classifier = new BackpropClassifier(trainingSet, 3, 5000, normalize = normalize)

      val accuracy = (for ((species, params) <- testSet) yield {
        classifier.predict(params) == species
      }).count { x: Boolean => x } / testSet.size.toDouble

      val time = (System.currentTimeMillis - start) / 1000.0
      logger.info(s"Fisher Iris backprop time: ${time}s")

      accuracy must be_>(0.8)  // 0.96 is typical
    }

    "classify the handwritten digits from the MNIST dataset" >> {
      val seed = Some(0L)
      val nHidden = 70
      val nRepeat = 1  // increase for better results
      val learnRate = 1.0f
      val expectedAccuracy = 0.9  // 0.95 with nRepeat=20

      val (trainingSet, testSet) = Mnist.trainingAndTestDataFloat()

      // Careful normalization is essential with this simple network.
      def scale(x: Float): Float = x / 256.0f // source byte range to (0; 1)
      val mean = trainingSet.flatMap(_._2).take(10000).map(scale).sum / 10000.0f
      def normalize(x: Float): Float = scale(x) - mean // balance around 0 for stability

      logger.info("creating the classifier")

      val start = System.currentTimeMillis

      val classifier = new BackpropClassifier(trainingSet, nHidden, nRepeat, learnRate, normalize, seed)

      logger.info("checking the accuracy")
      val accuracy = (for ((digit, params) <- testSet) yield {
        classifier.predict(params) == digit
      }).count { x: Boolean => x } / testSet.size.toDouble

      val time = (System.currentTimeMillis - start) / 1000.0
      logger.info(s"MNIST backprop time: ${time}s")

      accuracy must be_>(expectedAccuracy)
    }

    "use memory sparingly in predict" >> skipped {
      val vector = Seq.tabulate(100)(_.toFloat)
      val classifier = new BackpropClassifier(Seq((0, vector)), 10, 1)
      classifier.predict(vector) // allocate the fixed structures and train

      countAllocatedRepeat(10) {
        classifier.predict(vector)
      } must_== 0L  // passes
    }
  }
}

File: swiftlearner_cpp/test/nn/BackpropClassifierTest.cpp
#include "nn/BackpropClassifier.hpp"
#include "data/FisherIris.hpp"
#include "data/Mnist.hpp"
#include <gtest/gtest.h>
#include <iostream>
#include <chrono>

TEST(BackpropClassifierTest, ClassifyFisherIris) {
    // Load the Fisher Iris data
    auto [trainingSet, testSet] = FisherIris::loadData();

    // Normalize the input values to speed up learning
    auto normalize = [](float x) { return (x - 25.0f) / 25.0f; };

    auto start = std::chrono::high_resolution_clock::now();

    // Create the BackpropClassifier
    BackpropClassifier classifier(trainingSet, 3, 5000, 1.0f, normalize);

    // Calculate the accuracy
    double correctPredictions = 0;
    for (const auto& [species, params] : testSet) {
        if (classifier.predict(params) == species) {
            correctPredictions++;
        }
    }
    double accuracy = static_cast<double>(correctPredictions) / testSet.size();

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;

    std::cout << "Fisher Iris backprop time: " << duration << "s" << std::endl;
    std::cout << "Accuracy: " << accuracy << std::endl;

    // Assert that the accuracy is above the expected threshold
    ASSERT_GT(accuracy, 0.8); // 0.96 is typical
}

TEST(BackpropClassifierTest, ClassifyMnist) {
    int nHidden = 70;
    int nRepeat = 1;  // Increase for better results
    float learnRate = 1.0f;
    double expectedAccuracy = 0.87;  // 0.95 with nRepeat=20
    std::optional<long> seed = 0L;

    // Load the MNIST data
    auto [trainingSet, testSet] = Mnist::loadData(seed);

    // Careful normalization is essential with this simple network.
    auto scale = [](float x) { return x / 255.0f; }; // source byte range to (0; 1)

    //Calculate mean
    double mean = 0.0;
    int numSamples = 0;
    for (const auto& [digit, image] : trainingSet) {
      if (numSamples >= 10000) break; // Limit to 10000 *pixels*
        for (float pixelValue : image) {
            mean += scale(pixelValue);
            numSamples++;
            if (numSamples >= 10000) break;
        }
    }
    mean /= 10000.0;
    std::cout << "mean: " << mean << std::endl;

    auto normalize = [scale, mean](float x) { return scale(x) - mean; }; // balance around 0 for stability

    std::cout << "creating the classifier" << std::endl;

    auto start = std::chrono::high_resolution_clock::now();

    BackpropClassifier classifier(trainingSet, nHidden, nRepeat, learnRate, normalize, seed);

    std::cout << "checking the accuracy" << std::endl;
    double correctPredictions = 0;
    for (const auto& [digit, params] : testSet) {
        if (classifier.predict(params) == digit) {
            correctPredictions++;
        }
    }
    double accuracy = static_cast<double>(correctPredictions) / testSet.size();

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;

    std::cout << "MNIST backprop time: " << duration << "s" << std::endl;
    std::cout << "Accuracy: " << accuracy << std::endl;

    ASSERT_GT(accuracy, expectedAccuracy);
}